MSC2966: Usage of OAuth 2.0 Dynamic Client Registration in Matrix (#2966)
matrix-org/matrix-spec-proposals: proposals/2966-oauth2-dynamic-registration.md

Reviewer: Implementation requirements:
* [ ] Client using `web` flow
* [ ] Client using `native` flow
* [ ] Server supporting both flows (may be two different servers)

Reviewer: I haven't dug too far to find all the PRs or original implementations, but did find these points of evidence:
* https://github.com/matrix-org/matrix-js-sdk/pull/4070 (for Element Web via https://github.com/matrix-org/matrix-react-sdk/pull/12257 )
* https://github.com/element-hq/element-desktop/pull/1662
* https://github.com/matrix-org/matrix-rust-sdk/pull/2412
* The dendrite PR above

This MSC specifies what metadata values are required by the Matrix specification and how a client can register itself with a Matrix homeserver to get a client identifier.

The metadata names are registered in the IANA [OAuth Dynamic Client Registration Metadata](https://www.iana.org/assignments/oauth-parameters/oauth-parameters.xhtml#client-metadata) registry, and normative definitions of them are available in their respective RFCs in the registry.

#### Localizable metadata

#### `client_uri` and relationship with other URIs

The `client_uri` metadata is required to be a valid URI.
This URI must use the `https` scheme.

The host part of the URI must be a public hostname that is not a [public suffix](https://publicsuffix.org).
IP addresses and private hostnames like `localhost` are not allowed.

It is recommended that the `client_uri` is a web page that provides information about the client.
This page should be able to be accessed without requiring authentication.

This URI is a common base for all the other URIs in the metadata: those must be either on the same host or on a subdomain of the host of the `client_uri`.

> For example, if the `client_uri` is `https://example.com/`, then one of the `redirect_uris` can be `https://example.com/callback` or `https://app.example.com/callback`, but not `https://app.com/callback`.

Reviewer 2: For a self-hostable web client like Element web, the recommended uri would be `element.io` but if I host an instance of it on `example.com`, I wouldn't be able to set a proper `redirect_uri` with the recommended uri.
How important is this requirement really? What about informing the user about the differing host in the redirect uri during authorization?
Of course clients could also bundle in their landing pages and policies or set up a redirect to work around the issue.

Author: I think this is a feature, not a bug. "element.mycompany.corp" is "Element Web hosted by MyCompany", as there are no assurance that it's an untempered Element Web hosted by Element the company

Reviewer: Not for the MSC text, but Element Web could add an `/about` route that it ships for this purpose.

Reviewer: (considering thread as resolved - if this feels untrue, please open a new thread)

 - `policy_uri`: URL that points to a human-readable policy document for the client

All the URIs must use the `https` scheme and use the `client_uri` as a common base.

All of these metadata values are optional.

As per [RFC 7591 sec. 2.2](https://tools.ietf.org/html/rfc7591#section-2.2), these metadata values may be localized.
For example:

```json
{
  "client_name": "Digital mailbox",
  "client_name#en-US": "Digital mailbox",
  "client_name#en-GB": "Digital postbox",
  "client_name#fr": "Boîte aux lettres numérique",
  "tos_uri": "https://example.com/tos.html",
  "tos_uri#fr": "https://example.com/fr/tos.html",
  "policy_uri": "https://example.com/policy.html",
  "policy_uri#fr": "https://example.com/fr/policy.html"
}
```

#### Metadata values required by the OAuth 2.0 authorization grant flow

The following metadata values are required to be present to use the OAuth 2.0 authorization code grant, as described in [MSC2964]:

 - `redirect_uris`: Array of redirection URIs for use in redirect-based flows
 - `response_types`: Array of the OAuth 2.0 response types that the client may use
 - `grant_types`: Array of OAuth 2.0 grant types that the client may use

> To use this grant, along with refresh tokens:

Reviewer 2: This makes it sound like the items below are extra requirements to get a refresh token

```

```json
{
  "client_name": "My App",
  "client_name#fr": "Mon application",
  "client_uri": "https://example.com/",
  "logo_uri": "https://example.com/logo.png",
  "tos_uri": "https://example.com/tos.html",
  "tos_uri#fr": "https://example.com/fr/tos.html",
  "policy_uri": "https://example.com/policy.html",
  "policy_uri#fr": "https://example.com/fr/policy.html",
  "redirect_uris": ["https://app.example.com/callback"],
  "response_types": ["code"],
  "grant_types": ["authorization_code", "refresh_token"],
  "application_type": "web"
}
```

The server replies with a JSON object containing the `client_id` allocated, as well as all the metadata values that the server registered.

With the previous registration request, the server would reply with:

```json
{
  "client_id": "s6BhdRkqt3",
  "client_name": "My App",

  "client_uri": "https://example.com/",
  "logo_uri": "https://example.com/logo.png",
  "tos_uri": "https://example.com/tos.html",
  "policy_uri": "https://example.com/policy.html",
  "redirect_uris": ["https://app.example.com/callback"],
  "response_types": ["code"],
  "grant_types": ["authorization_code", "refresh_token"],
  "application_type": "web"
}
```

> **Note**: in this example, the server has not registered the locale-specific values for `client_name`, `tos_uri`, and `policy_uri`, which is why they are not present in the response.

Reviewer 2: Maybe it'd be useful to add examples which show that the server could ignore any values, not just localized ones (except required values specified in this MSC) or choose to modify values. For example a homeserver could ignore the `grant_types` sent by the client and register that it has access to any supported `grant_types` for a simpler implementation (at least according to RFC 7591)

Author: I adapted the example and stated that the server must ignore unknown metadata/values in 4deb3c85

# MSC2966: Usage of OAuth 2.0 Dynamic Client Registration in Matrix

This proposal is part of the broader [MSC3861: Next-generation auth for Matrix, based on OAuth 2.0/OIDC](https://github.com/matrix-org/matrix-spec-proposals/pull/3861).

This MSC specifies how Matrix clients SHOULD leverage the OAuth 2.0 Dynamic Client Registration Protocol ([RFC 7591](https://tools.ietf.org/html/rfc7591)) to register themselves before initiating an authorization flow.

> 

Author: That's a good suggestion, thanks!

## Proposal

### Prerequisites

This proposal requires the client to know the following authorization server metadata about the homeserver:

 - `registration_endpoint`: the URL where the client is able to register itself.

The discovery of the above metadata is out of scope for this MSC and is currently covered by [MSC2965](https://github.com/matrix-org/matrix-doc/pull/2965).

### Client metadata

In OAuth 2.0, clients have a set of metadata values associated with their client identifier at an authorization server.
These values are used to describe the client to the user and define how the client interacts with the authorization server.

This MSC specifies what metadata values are required by the Matrix specification and how a client can register itself with a Matrix homeserver to get a client identifier.

The metadata names are registered in the IANA [OAuth Dynamic Client Registration Metadata](https://www.iana.org/assignments/oauth-parameters/oauth-parameters.xhtml#client-metadata) registry, and normative definitions of them are available in their respective RFCs in the registry.

#### Localizable metadata

> 

Author: Woops, thanks! Fixed in fae9c291

#### `client_uri` and relationship with other URIs

Per [RFC 7591](https://tools.ietf.org/html/rfc7591), the `client_uri` MUST be a valid URL that SHOULD give the user more information about the client.

> This URL SHOULD NOT require authentication to access.

Reviewer 3: is `client_uri` required? 

If not, some "If present" wording would be helpful, and explain how the constraints on the other URIs work when it is absent.

If it is required, call it out explicitly, and maybe say what should happen if it's not there?

Author: Clarified that it is required in `f3d30e96`

 - `redirect_uris`: Array of redirection URIs for use in redirect-based flows
 - `response_types`: Array of the OAuth 2.0 response types that the client may use
 - `grant_types`: Array of OAuth 2.0 grant types that the client may use
 - `token_endpoint_auth_method`: String indicator of the requested authentication method for the token endpoint

The homeserver MUST support the `none` value for the `token_endpoint_auth_method`, as most Matrix clients are client-side only, do not have a server component, and therefore are public clients.

To use this grant:

 - the `redirect_uris` MUST have at least one value
 - the `response_types` MUST include `code`
 - the `grant_types` MUST include `authorization_code` and `refresh_token`

#### Redirect URI validation

The redirect URI plays a critical role in validating the authenticity of the client.
The client 'proves' its identity by demonstrating that it controls the redirect URI.
This is why it is critical to have strict validation of the redirect URI.

The `application_type` metadata is used to determine the type of client.
It defaults to `web` if not present, and can be set to `native` to indicate that the client is a native application.

> In all cases, the redirect URI MUST not have a fragment component.

Reviewer 3: ```suggestion
In all cases, the redirect URI MUST NOT have a fragment component.
```

#### Web clients

`web` clients can use redirect URIs that:

 - MUST use the `https` scheme
 - MUST omit the port (to use the default port for https: 443)
 - MUST not use a user or password in the authority component of the URI
 - MUST use the client URI as a common base for the authority component

Examples of valid redirect URIs (with `https://example.com/` as the client URI):

 - `https://example.com/callback`
 - `https://app.example.com/callback`
 - `https://example.com/?query=value`

Examples of invalid redirect URIs (with `https://example.com/` as the client URI):

 - `https://example.com/callback#fragment`
 - `https://example.com:8080/callback`
 - `http://example.com/callback`
 - `http://localhost/`

#### Native clients

`native` clients can use three types of redirect URIs:

 1. Private-Use URI Scheme:
    - the scheme MUST be prefixed with the client URI hostname in reverse-DNS notation. For example, if the client URI is `https://example.com/`, then a valid custom URI scheme would be `com.example.app:/`.

>     - the URI MUST not have an authority component. This means that it MUST have either a single slash or none immediately following the scheme, with no hostname, username, or port.

Reviewer 3: ```suggestion
    - the URI MUST NOT have an authority component. This means that it MUST have either a single slash or none immediately following the scheme, with no hostname, username, or port.
```

 2. "http" URIs on the loopback interface:
    - it MUST use the `http` scheme
    - the host part MUST be `localhost`, `127.0.0.1`, or `[::1]`

>     - it MUST have no port registered. The homeserver MUST then accept any port number during the authorization flow.

Reviewer 3: ```suggestion
    - it (the registered redirect URI) MUST NOT contain a port number. The homeserver MUST then accept any port number during the authorization flow.
```

Author: a98b2ba1

 3. Claimed "https" Scheme URI:
    - some operating systems allow apps to claim "https" scheme URIs in the domains they control
    - when the browser encounters a claimed URI, instead of the page being loaded in the browser, the native app is launched with the URI supplied as a launch parameter
    - the same rules as for `web` clients apply

These restrictions are the same as defined by [RFC8252 sec. 7](https://tools.ietf.org/html/rfc8252#section-7).

Examples of valid redirect URIs (with `https://example.com/` as the client URI):

 - `com.example.app:/callback`
 - `com.example:/`
 - `com.example:callback`
 - `http://localhost/callback`
 - `http://127.0.0.1/callback`
 - `http://[::1]/callback`

Examples of invalid redirect URIs (with `https://example.com/` as the client URI):

 - `example:/callback`
 - `com.example.app://callback`
 - `https://localhost/callback`

>  - `http://localhost:1234/callback`

Author: Oh, those were supposed to be sub-sections of `Redirect URI validation`. Fixed in 1768d54f

To register, the client sends an HTTP POST to the `registration_endpoint` with its metadata as JSON in the body.
For example, the client could send the following registration request:

```http
POST /register HTTP/1.1
Content-Type: application/json
Accept: application/json
Server: auth.example.com
```

```json
{
  "client_name": "My App",
  "client_name#fr": "Mon application",
  "client_uri": "https://example.com/",
  "logo_uri": "https://example.com/logo.png",
  "tos_uri": "https://example.com/tos.html",
  "tos_uri#fr": "https://example.com/fr/tos.html",
  "policy_uri": "https://example.com/policy.html",
  "policy_uri#fr": "https://example.com/fr/policy.html",
  "redirect_uris": ["https://app.example.com/callback"],
  "token_endpoint_auth_method": "none",
  "response_types": ["code"],
  "grant_types": [
    "authorization_code",
    "refresh_token",
    "urn:ietf:params:oauth:grant-type:token-exchange"
  ],
  "application_type": "web"
}
```

The server replies with a JSON object containing the `client_id` allocated, as well as all the metadata values that the server registered.

> It MUST ignore fields, `grant_types` and `response_types` that are not understood by the server.

Author: Clarified that in 6e32a716

With the previous registration request, the server would reply with:

```json
{
  "client_id": "s6BhdRkqt3",
  "client_name": "My App",
  "client_uri": "https://example.com/",
  "logo_uri": "https://example.com/logo.png",
  "tos_uri": "https://example.com/tos.html",
  "policy_uri": "https://example.com/policy.html",
  "redirect_uris": ["https://app.example.com/callback"],
  "response_types": ["code"],
  "grant_types": ["authorization_code", "refresh_token"],
  "application_type": "web"
}

> ```

Author: As defined in RFC7591, the client doesn't have a client_secret because the `token_endpoint_auth_method` is set to `none`. It was missing from the response here though, so I added it in 650a4498

**Note**: in this example, the server has not registered the locale-specific values for `client_name`, `tos_uri`, and `policy_uri`, which is why they are not present in the response. The server also does not support the `urn:ietf:params:oauth:grant-type:token-exchange` grant type, which is why it is not present in the response.

The client MUST store the `client_id` for future use.

> It SHOULD reuse the `client_id` for all future authorization requests done against the same homeserver.

Reviewer 4: So to be clear, the client should associate the `client_id` to the server name of the homeserver? The `issuer` field from [RFC8414](https://datatracker.ietf.org/doc/html/rfc8414) is ignored, since it doesn't appear to be mentioned in any MSC?

Won't that complicate the migration of current implementations that should have associated their `client_id` to the `issuer`?

Reviewer 3: @zecakeh any further comments on this thread?

Reviewer 4: No, thanks.

> It SHOULD reuse the `client_id` for all future authorization requests done against the same homeserver.

Author: Makes sense, I think the reason I did it this way is so that the fields in the request example make more sense if they are defined before?

> It SHOULD reuse the `client_id` for all future authorization requests done against the same homeserver.

Reviewer 3: I think it's fine, and normal, for an example to reference fields that haven't been defined yet.

Again, not insisting on a change

## Potential issues

Because each client on each user device will do its own registration, they will all have different `client_id`s.
This means that the number of client registrations will grow exponentially.
A subsequent MSC could be proposed to identify multiple instances of the same client using signed client metadata.

## Alternatives

An alternative approach would be to have the client host a JSON file containing its metadata and use that URL as the `client_id`.

> This is what the following [*OAuth Client ID Metadata Document* draft](https://datatracker.ietf.org/doc/html/draft-parecki-oauth-client-id-metadata-document) proposes.

Reviewer 3: ```suggestion
This is what the [*OAuth Client ID Metadata Document* draft](https://datatracker.ietf.org/doc/html/draft-parecki-oauth-client-id-metadata-document) proposes.
```

Reviewer 3: or:

```suggestion
This is what the following document proposes: [*OAuth Client ID Metadata Document* draft](https://datatracker.ietf.org/doc/html/draft-parecki-oauth-client-id-metadata-document).
```

## Proposal

### Prerequisites

This proposal requires the client to know the following authorization server metadata about the homeserver:

 - `registration_endpoint`: the URL where the client is able to register itself.

The discovery of the above metadata is out of scope for this MSC and is currently covered by [MSC2965](https://github.com/matrix-org/matrix-doc/pull/2965).

### Client metadata

In OAuth 2.0, clients have a set of metadata values associated with their client identifier at an authorization server.
These values are used to describe the client to the user and define how the client interacts with the authorization server.

This MSC specifies what metadata values are required by the Matrix specification and how a client can register itself with a Matrix homeserver to get a client identifier.

The metadata names are registered in the IANA [OAuth Dynamic Client Registration Metadata](https://www.iana.org/assignments/oauth-parameters/oauth-parameters.xhtml#client-metadata) registry, and normative definitions of them are available in their respective RFCs in the registry.

#### `client_uri` and relationship with other URIs

> Per [RFC 7591](https://tools.ietf.org/html/rfc7591), the `client_uri` MUST be a valid URL that SHOULD give the user more information about the client.

Reviewer 5: I'm unsure about requiring a client URI, as it means authors of all clients *must* have a website of some form. Given e.g. `client_name` is optional, what's the rationale of requiring `client_uri`?

Some homeservers may wish to try and lock down which clients can use them, but I think that is a special case.

Author: The whole reason of this section is to mitigate client impersonation as much as possible.

A web client claiming to be 'Element' will not be able to use `https://element.io/` as client URI unless they can host their client in a subdomain of element.io. We display that client_uri to the end user, so that they can have some trust with it, without displaying the (potentially ugly and more technical) redirect_uri to build that same trust.

It does not mean having to have a proper dedicated domain name, free/shared hosting services will work. MAS will use the public suffix list to know whether a client can claim a domain or not, to allow for example `https://myclient.github.io/` (but not `https://github.io/`) as `client_uri`

Of course this cannot be bulletproof, as described in the 'security considerations' section

This URL SHOULD NOT require authentication to access.

The `client_uri` is required and the server MAY reject client registrations with an invalid or missing `client_uri`.
This URI is a common base for all the other URIs in the metadata: those MUST be either on the same host or on a subdomain of the host of the `client_uri`.

> For example, if the `client_uri` is `https://example.com/`, then one of the `redirect_uris` can be `https://example.com/callback` or `https://app.example.com/callback`, but not `https://app.com/callback`.

Reviewer 6: Also seems like no consideration has gone into protocol & port components of the URI, do those have to match too?

> For example, if the `client_uri` is `https://example.com/`, then one of the `redirect_uris` can be `https://example.com/callback` or `https://app.example.com/callback`, but not `https://app.com/callback`.

Reviewer 6: If the URI is on a child subdomain does the path matter? Does the path matter at all? It says "common base" and that seems to imply caring about path but I don't see any examples around that

> For example, if the `client_uri` is `https://example.com/`, then one of the `redirect_uris` can be `https://example.com/callback` or `https://app.example.com/callback`, but not `https://app.com/callback`.

Author: This is an attempt to mitigate as much as possible client impersonation. This is so that I cannot claim to be 'https://element.io/' if my redirect URI is 'https://evil.com/callback'

This is what the RFC says in the security considerations:

>   In a situation where the authorization server is supporting open
   client registration, it must be extremely careful with any URL
   provided by the client that will be displayed to the user (e.g.,
   "logo_uri", "tos_uri", "client_uri", and "policy_uri").  For
   instance, a rogue client could specify a registration request with a
   reference to a drive-by download in the "policy_uri", enticing the
   user to click on it during the authorization.  The authorization
   server SHOULD check to see if the "logo_uri", "tos_uri",
   "client_uri", and "policy_uri" have the same host and scheme as the
   those defined in the array of "redirect_uris" and that all of these
   URIs resolve to valid web pages.

On the scheme, I think we should clarify that all the URIs (except the redirect_uris) must be `https`, should be pointing to HTML web pages (except for the logo_uri)

On the port, I think we can relax the requirement and allow any port, and not require to match between the URIs

> For example, if the `client_uri` is `https://example.com/`, then one of the `redirect_uris` can be `https://example.com/callback` or `https://app.example.com/callback`, but not `https://app.com/callback`.

Reviewer 6: > I think we should clarify that all the URIs (except the redirect_uris) must be https, should be pointing to HTML web pages (except for the logo_uri)

So TOS & Policy can't be PDFs or other document formats? That seems like an odd restriction to state in the Matrix spec.

> For example, if the `client_uri` is `https://example.com/`, then one of the `redirect_uris` can be `https://example.com/callback` or `https://app.example.com/callback`, but not `https://app.com/callback`.

Reviewer 6: Shouldn't we also limit then the format allowed on `logo_uri` so it is something that the likes of MAS is capable of rendering? Or are CR3 files fine?

> For example, if the `client_uri` is `https://example.com/`, then one of the `redirect_uris` can be `https://example.com/callback` or `https://app.example.com/callback`, but not `https://app.com/callback`.

Reviewer 6: > On the port, I think we can relax the requirement and allow any port, and not require to match between the URIs

Note to self, the fixup I just landed in js-sdk does not forbid port but does check that they match

> For example, if the `client_uri` is `https://example.com/`, then one of the `redirect_uris` can be `https://example.com/callback` or `https://app.example.com/callback`, but not `https://app.com/callback`.

Reviewer 6: > So TOS & Policy documents can't be PDFs or other document formats? That seems like an odd restriction to state in the Matrix spec.

Seems like Element already uses PDFs as privacy docs e.g. for IS: https://static.element.io/legal/identity-server-privacy-policy.pdf

> For example, if the `client_uri` is `https://example.com/`, then one of the `redirect_uris` can be `https://example.com/callback` or `https://app.example.com/callback`, but not `https://app.com/callback`.

Author: I have relaxed the requirement on the port in 03ba6f2d

#### User-visible metadata values

The following metadata values SHOULD be used by clients to help users identify the client:

 - `client_name`: Human-readable name of the client to be presented to the user
 - `logo_uri`: URL that references a logo for the client
 - `tos_uri`: URL that points to a human-readable terms of service document for the client
 - `policy_uri`: URL that points to a human-readable policy document for the client

All the URIs MUST use the `https` scheme and use the `client_uri` as a common base.

> If provided by the client, the homeserver SHOULD show or link to the `tos_uri` and `policy_uri` to the user.

Author: Good suggestion, did that in 8b2f387b

 - `grant_types`: Array of OAuth 2.0 grant types that the client may use
 - `token_endpoint_auth_method`: String indicator of the requested authentication method for the token endpoint

The homeserver MUST support the `none` value for the `token_endpoint_auth_method`, as most Matrix clients are client-side only, do not have a server component, and therefore are public clients.

To use this grant:

 - the `redirect_uris` MUST have at least one value
 - the `response_types` MUST include `code`
 - the `grant_types` MUST include `authorization_code` and `refresh_token`

#### Redirect URI validation

The redirect URI plays a critical role in validating the authenticity of the client.
The client 'proves' its identity by demonstrating that it controls the redirect URI.
This is why it is critical to have strict validation of the redirect URI.

The `application_type` metadata is used to determine the type of client.
It defaults to `web` if not present, and can be set to `native` to indicate that the client is a native application.

In all cases, the redirect URI MUST NOT have a fragment component.

##### Web clients

`web` clients can use redirect URIs that:

 - MUST use the `https` scheme

>  - MUST omit the port (to use the default port for https: 443)

Reviewer 5: Why can't we use redirect URLs on different ports?

Reviewer 6: Yeah agreed, this breaks the ability to test local development builds running in vite server, webpack dev server etc, especially if user lacks root access.

Also makes it harder to do CI end to end tests where we may need to use multiple ports at once to parallelise.

Author: Relaxed in 03ba6f2d4db81a8e79b286254365f6230eeb7817

With the previous registration request, the server would reply with:

```json
{
  "client_id": "s6BhdRkqt3",
  "client_name": "My App",
  "client_uri": "https://example.com/",
  "logo_uri": "https://example.com/logo.png",
  "tos_uri": "https://example.com/tos.html",
  "policy_uri": "https://example.com/policy.html",
  "redirect_uris": ["https://app.example.com/callback"],
  "response_types": ["code"],
  "grant_types": ["authorization_code", "refresh_token"],
  "application_type": "web"
}
```

**Note**: in this example, the server has not registered the locale-specific values for `client_name`, `tos_uri`, and `policy_uri`, which is why they are not present in the response. The server also does not support the `urn:ietf:params:oauth:grant-type:token-exchange` grant type, which is why it is not present in the response.

The client MUST store the `client_id` for future use.
It SHOULD reuse the `client_id` for all future authorization requests done against the same homeserver.

## Potential issues

Because each client on each user device will do its own registration, they will all have different `client_id`s.

> This means that the number of client registrations will grow exponentially.

Reviewer 5: Yeah, I think we should allow clearing up of registered clients that are not used by any sessions?

I think to do so we'd need some response type when the client tries and uses a deleted `client_id`?

> This means that the number of client registrations will grow exponentially.

Author: 'Exponentially' is indeed wrong here. Fixed that in 6b03a534

Problem is, with the current proposal, we don't really have a way to notice that until we redirected to the homeserver, and we usually don't want to redirect back to the client if we don't trust the redirect_uri provided…

If we wanted to implement 'garbage collection' of client IDs, we should use [RFC9126 pushed access requests](https://datatracker.ietf.org/doc/html/rfc9126), so that we can give feedback that the client_id is invalid right away, before the redirect.

I would leave that as a future improvement though, I'm not too concern about the number of registration completely exploding

> This means that the number of client registrations will grow exponentially.

Author: I've added words on how to mitigate in b6abea5f

> This means that the number of client registrations will grow exponentially.

Reviewer 2: One possible way to deal with this would be to implement the client read request from RFC 7592. The suggestion in https://github.com/matrix-org/matrix-spec-proposals/commit/b6abea5f854d7d5e00bc47b93422e9ca62f71058 doesn't really clean up much.

> This means that the number of client registrations will grow exponentially.

Author: It still means you can clean up any user <-> client relationship, and just have to retain client ID and the redirect URIs, instead of all the metadata fields.

Problem with client read request, is that it makes us implement a relatively big new mechanism (with only the read request, not the update and delete), including issuing client registration tokens, to fix a potential problem that we yet don't have (as we don't yet expire client IDs)

> This means that the number of client registrations will grow exponentially.

Reviewer 5: One thing I don't really follow is: clients presumably can't assume that they are registered on a given homeserver, so what's the difference for a client between encountering a new homeserver and a homeserver that has deleted their registration (assuming the homeserver has no active sessions with the client ID)?

> This means that the number of client registrations will grow exponentially.

Author: > clients presumably can't assume that they are registered on a given homeserver

They do that through local storage. When the client does the registration, it can save the client_id in a `hs => client_id` mapping. It does mean also that right now each instance of a client on each device is considered as a different client

> This means that the number of client registrations will grow exponentially.

Reviewer 5: Yeah, but generally clients will have to go through the registration flow. At which point, can't we just pretty much always do the client registration (and deduplicate server side)? That way we can easily delete client IDs.

My concern with not addressing this now is that its going to be hard to add the ability to delete client IDs in a backwards compatible way in the future, if they're assuming they can reuse any previous registration

> This means that the number of client registrations will grow exponentially.

Author: I've suggested deduplication, plus a long-term suggestion to avoid the 'we cannot delete client IDs' problem: a0e8b376

> This means that the number of client registrations will grow exponentially.

Reviewer 3: Unfortunately I don't think that solves the problem right now.

(Some members of) the SCT remain concerned that, without a mechanism in the MSC by which servers can safely garbage-collect old client IDs, it's going to be a real pain to fix in the future.

One suggestion was 

> can we just specify that `400 M_INVALID_CLIENT_ID` upon first use means "please re-register/I don't know where you got that from, but it's wrong"

> This means that the number of client registrations will grow exponentially.

Author: If the inability to GC old clients is a hard blocker, we need to solve that.

The core of the issue is that the client cannot poke whether their client ID is valid or not. When they start using it, it's already too late, because they have started a auth grant, which implies a browser redirect.

Whenever there is an error in the flow (like the user denied access), we usually redirect back to the client with an `error`. Problem is, to redirect back to the client, you have to know the client, which isn't the case if you garbage-collected it…

So, **if we want the client to be able to poke whether the `client_id` is still valid** we either need to:

 - implement the [client read request from RFC7592](https://datatracker.ietf.org/doc/html/rfc7592#section-2.1). This way the client has a way to get its own registered metadata. Drawback of this is one extra call, and one extra token to keep track of for the client with little value, and we would only partially implement RFC7592, as the other parts (update & delete) wouldn't make sense for us.
 - change MSC2966 to use [RFC 9126 Pushed Auth Requests](https://datatracker.ietf.org/doc/html/rfc9126), which would make the client prepare the auth request through an HTTP call instead of a redirect, which means we can give it an error properly here

Long term, I would like to have pushed auth requests. But this would add an extra step for clients to login, and significant changes in the implementations now if we want to unblock the MSC.

**Alternatively**, what if we said something among the lines of

> The homeserver may delete client registrations that had no active session and have not been used in more than a month time.
> Therefore, clients must do a new client registration if they want to authenticate against a homeserver they haven't used in a while.

To be clear, I don't like the idea of putting an arbitrary date in there, but this is the easiest way we can get out of it without changing the implementations

> This means that the number of client registrations will grow exponentially.

Reviewer 3: > Whenever there is an error in the flow (like the user denied access), we usually redirect back to the client with an `error`. Problem is, to redirect back to the client, you have to know the client, which isn't the case if you garbage-collected it…

I think I (and maybe other members of the SCT) am slightly failing to join the dots about how an expired client_id is a problem, so to set this out more explicitly, it would be something like:

 1. Client knows it has previously registered and has a `client_id` which it believes to be valid
 2. It therefore *redirects* the user to the authorization endpoint, per [MSC2964](https://github.com/sandhose/matrix-spec-proposals/blob/msc/sandhose/oauth2-profile/proposals/2964-oauth2-profile.md#sample-authorization-request)
 3. The server doesn't recognise the `client_id` so needs to return an error

Obviously, a `400 M_INVALID_CLIENT_ID` doesn't help us here, but I'm not quite following why the server cannot return an error to the client via the normal callback mechanism? It seems like the server has everything it needs for this?

> This means that the number of client registrations will grow exponentially.

Author: This is because OAuth 2.0 wants to prevent your server becoming an open redirector.

[RFC 6749 section 4.1.2.1](https://www.rfc-editor.org/rfc/rfc6749#section-4.1.2.1) states

>    If the request fails due to a missing, invalid, or mismatching
   redirection URI, or if the client identifier is missing or invalid,
   the authorization server SHOULD inform the resource owner of the
   error and MUST NOT automatically redirect the user-agent to the
   invalid redirection URI.

In this context, 'resource owner' is the end user, 'user-agent' is the web browser; so they tell you that you should display an error page to the user, but not go back to the client.

This is also expanded a bit in [RFC 9700 section 4.11.2](https://datatracker.ietf.org/doc/html/rfc9700#section-4.11.2)

> This means that the number of client registrations will grow exponentially.

Reviewer 7: > The homeserver may delete client registrations that had no active session and have not been used in more than a month time.
> Therefore, clients must do a new client registration if they want to authenticate against a homeserver they haven't used in a while.

This option sounds the easiest. Does it even need to be such a long time? Servers could just check for client IDs with no valid access tokens and clean them up if they're more than a day old for example. Then just define that clients must not store client IDs after logout and must re-register if the user spends more than a few hours on the pre-login screen or something like that

> This means that the number of client registrations will grow exponentially.

Author: Did that in 1f489eee

> This means that the number of client registrations will grow exponentially.

Reviewer 3: looks better to me now

> A subsequent MSC could be proposed to identify multiple instances of the same client using signed client metadata.

Reviewer 5: Presumably the server can deduplicate clients that register with the exact same JSON?

Author: This is an implementation detail but yes.

Signed client metadata would potentially allow changing this metadata over time as needed

Reviewer 8: Shouldn't de-duplication also be suggested as a potential mitigation technique in this section; similar to suggesting deletion inactive client registrations?

Or are you trying to discourage implementations from going down that route?

Author: I suggested deduplication in a0e8b376

I also implemented that dedupe in MAS: https://github.com/element-hq/matrix-authentication-service/pull/4293

## Alternatives

An alternative approach would be to have the client host a JSON file containing its metadata and use that URL as the `client_id`.
This is what the [*OAuth Client ID Metadata Document* draft](https://datatracker.ietf.org/doc/html/draft-parecki-oauth-client-id-metadata-document) proposes.

> This approach has the advantage of being able to use the same `client_id` for different instances of the same client, but it has the disadvantage of requiring the client to host a JSON file on its own domain, as well as difficulties in handling updates to the metadata.

Reviewer 5: If `client_uri` is mandatory, then hosting a JSON file isn't really a big concern no? Though personally I think we shouldn't require a `client_uri` if we don't need to.

Author: I think this has the advantage of not requiring any communication from the homeserver to the 'client_uri'. This is interesting for example during development where my `https://local-dev.mycorp` might resolve to my local machine, but isn't reachable by the homeserver.

It also introduces a whole level of questions like:

 - should the HS cache the client metadata fetched?
 - what if the metadata changes?
 - how does the client manage rollout of new versions, with intentional incompatible changes in the metadata?

Those are better handled if the client is the one pushing the metadata, instead of the homeserver pulling it

## Security considerations

The restrictions on the metadata values laid out in this MSC are a best effort to prevent client impersonation, but they are not flawless.

For web clients, it relies on the client's ability to prove ownership of the redirect URI, which can be guaranteed to some extent by sane DNS management and its use of TLS.
If a client-related domain name hosts an open redirector, it could be used to impersonate the client.

For native clients, because they can use private-use URI schemes and localhost redirectors, it relies more on the underlying operating system's security model and their application distribution model.

> A good example of this is if a mobile client distributed through an app store registers the `app.acme.corp:` scheme in an effort to impersonate "ACME Corp's" app, then "ACME Corp" would have a valid case to take down the malicious app from the app store.

Reviewer 5: This doesn't protect against clients using localhost schemes? Or against clients that have embedded browsers that can intercept any scheme? Doesn't android allow you to intercept any/all URL loads for embedded web views?

TBH, I'm wondering whether this whole section should just be "the server cannot trust that the client is who they say they are".

Author: `localhost` schemes won't work in many cases, including mobile apps, and some environments might [require specific entitlements](https://developer.apple.com/documentation/BundleResources/Entitlements/com.apple.security.network.server).

In short,

 - web apps are protected against impersonation
 - mobile native apps have some protection against impersonation (and a strong case to take down apps from the app store if impersonation is detected)
 - desktop native apps is the area where we unfortunately can't do as much

# MSC2966: Usage of OAuth 2.0 Dynamic Client Registration in Matrix

This proposal is part of the broader [MSC3861: Next-generation auth for Matrix, based on OAuth 2.0/OIDC](https://github.com/matrix-org/matrix-spec-proposals/pull/3861).

This MSC specifies how Matrix clients should leverage the OAuth 2.0 Dynamic Client Registration Protocol ([RFC 7591](https://tools.ietf.org/html/rfc7591)) to register themselves before initiating an authorization flow.

## Proposal

### Prerequisites

This proposal requires the client to know the following authorization server metadata about the homeserver:

 - `registration_endpoint`: the URL where the client is able to register itself.

> The discovery of the above metadata is out of scope for this MSC and is currently covered by [MSC2965](https://github.com/matrix-org/matrix-doc/pull/2965).

Reviewer: presumably if there's no `registration_endpoint`, the server doesn't support dynamic client registration. That should be documented in the MSC as a thing clients should expect.

Author: I'm not sure how to handle this. Basically there are two possibilities:

 - make client registration mandatory. That would be aligned with the 'openness' of Matrix, as you should be able to use any client with any compliant server. Not supporting this would then mean the server not spec-compliant?
 - make it optional and explain here what this means. This makes it clearer in the spec that its possible to use this in 'closed' ecosystems, where servers impose which client is allowed to use it

Reviewer: hmm, looks like MSC2965 marks it required (and I just missed that), so would definitely lean towards openness. Something we may need to clarify in the spec PR (rather than MSC) is that "registration" is just so the client can talk to the server, not for creating new accounts. The client may still be unable to create new accounts or log in to existing accounts later in the process.

### Client metadata

In OAuth 2.0, clients have a set of metadata values associated with their client identifier at an authorization server.
These values are used to describe the client to the user and define how the client interacts with the authorization server.

This MSC specifies what metadata values are required by the Matrix specification and how a client can register itself with a Matrix homeserver to get a client identifier.

The metadata names are registered in the IANA [OAuth Dynamic Client Registration Metadata](https://www.iana.org/assignments/oauth-parameters/oauth-parameters.xhtml#client-metadata) registry, and normative definitions of them are available in their respective RFCs in the registry.

#### Localizable metadata

#### `client_uri` and relationship with other URIs

> The `client_uri` metadata is required to be a valid URI.
> This URI must use the `https` scheme.
> 
> The host part of the URI must be a public hostname that is not a [public suffix](https://publicsuffix.org).
> IP addresses and private hostnames like `localhost` are not allowed.
> 
> It is recommended that the `client_uri` is a web page that provides information about the client.
> This page should be able to be accessed without requiring authentication.

Reviewer: I believe this is all covered by the RFC:

```suggestion
Per [RFC 7591](https://tools.ietf.org/html/rfc7591), the `client_uri` MUST be a valid URL
that SHOULD give the user more information about the client. This URL SHOULD NOT
require authentication to access.
```

This URI is a common base for all the other URIs in the metadata: those must be either on the same host or on a subdomain of the host of the `client_uri`.
For example, if the `client_uri` is `https://example.com/`, then one of the `redirect_uris` can be `https://example.com/callback` or `https://app.example.com/callback`, but not `https://app.com/callback`.

#### User-visible metadata values

The following metadata values should be used by clients to help users identify the client:

 - `client_name`: Human-readable name of the client to be presented to the user
 - `logo_uri`: URL that references a logo for the client
 - `tos_uri`: URL that points to a human-readable terms of service document for the client

>  - `policy_uri`: URL that points to a human-readable policy document for the client

Reviewer: blocking: what is the difference between policy and terms of service?

Reviewer: ah, alright. If they're from the RFC then let's just copy them.

All of these metadata values are optional.

As per [RFC 7591 sec. 2.2](https://tools.ietf.org/html/rfc7591#section-2.2), these metadata values may be localized.
For example:

```json
{
  "client_name": "Digital mailbox",
  "client_name#en-US": "Digital mailbox",
  "client_name#en-GB": "Digital postbox",
  "client_name#fr": "Boîte aux lettres numérique",
  "tos_uri": "https://example.com/tos.html",
  "tos_uri#fr": "https://example.com/fr/tos.html",
  "policy_uri": "https://example.com/policy.html",
  "policy_uri#fr": "https://example.com/fr/policy.html"
}
```

#### Metadata values required by the OAuth 2.0 authorization grant flow

The following metadata values are required to be present to use the OAuth 2.0 authorization code grant and refresh token grant, as described in [MSC2964]:

 - `redirect_uris`: Array of redirection URIs for use in redirect-based flows
 - `response_types`: Array of the OAuth 2.0 response types that the client may use
 - `grant_types`: Array of OAuth 2.0 grant types that the client may use
 - `token_endpoint_auth_method`: String indicator of the requested authentication method for the token endpoint

> The homeserver must support the `none` value for the `token_endpoint_auth_method`, as most Matrix clients are client-side only, do not have a server component, and therefore are public clients.

Reviewer: ```suggestion
The homeserver MUST support the `none` value for the `token_endpoint_auth_method`, as most Matrix clients are client-side only, do not have a server component, and therefore are public clients.
```

To use this grant:

>  - the `redirect_uris` must have at least one value
>  - the `response_types` must include `code`
>  - the `grant_types` must include `authorization_code` and `refresh_token`

Reviewer: ```suggestion
 - the `redirect_uris` MUST have at least one value
 - the `response_types` MUST include `code`
 - the `grant_types` MUST include `authorization_code` and `refresh_token`
```

    "refresh_token",
    "urn:ietf:params:oauth:grant-type:token-exchange"
  ],
  "application_type": "web"
}
```

The server replies with a JSON object containing the `client_id` allocated, as well as all the metadata values that the server registered.
It must ignore fields, `grant_types` and `response_types` that are not understood by the server.

With the previous registration request, the server would reply with:

```json
{
  "client_id": "s6BhdRkqt3",
  "client_name": "My App",
  "client_uri": "https://example.com/",
  "logo_uri": "https://example.com/logo.png",
  "tos_uri": "https://example.com/tos.html",
  "policy_uri": "https://example.com/policy.html",
  "redirect_uris": ["https://app.example.com/callback"],
  "response_types": ["code"],
  "grant_types": ["authorization_code", "refresh_token"],
  "application_type": "web"
}
```

**Note**: in this example, the server has not registered the locale-specific values for `client_name`, `tos_uri`, and `policy_uri`, which is why they are not present in the response. The server also does not support the `urn:ietf:params:oauth:grant-type:token-exchange` grant type, which is why it is not present in the response.

> The client must store the `client_id` for future use.

Reviewer: Mentioned in the umbrella MSC: does the client need to remember this over multiple browsers/sessions, or just while the single user session logs in?

# MSC2966: Usage of OAuth 2.0 Dynamic Client Registration in Matrix

This proposal is part of the broader [MSC3861: Next-generation auth for Matrix, based on OAuth 2.0/OIDC](https://github.com/matrix-org/matrix-spec-proposals/pull/3861).

This MSC specifies how Matrix clients SHOULD leverage the OAuth 2.0 Dynamic Client Registration Protocol ([RFC 7591](https://tools.ietf.org/html/rfc7591)) to register themselves before initiating an authorization flow.

In brief, once a client has obtained the homeserver's OAuth 2.0 metadata per [MSC2965](https://github.com/matrix-org/matrix-spec-proposals/pull/2965), and before it can initiate an authorization request per [MSC2964](https://github.com/matrix-org/matrix-spec-proposals/pull/2964), the client must register itself with that homeserver to obtain a `client_id`.

> It stores this `client_id` and uses it for all future authorization requests against the homeserver.

Reviewer 9: This is probably a stupid question, but given i'm thinking it, others reading this probably will too: What's the difference between the `client_id` and the `device_id` for a given session, and is there any recommendation/expectation for them to be the same or not?

I assume we assume them to be different here because:
 a) the client_id is generated by the IdP rather than the HS, and we don't want the IdP to have to get a device ID from the HS?
 b) currently clients can specify their own device IDs
 c) clients only need a device ID if they're speaking E2EE, and the client might not implement E2EE?

Is this right, and what else am i missing? It feels like it's worth spelling out here for idiots like me :)

Reviewer 6: A `client_id` isn't unique to that device, it is likely (the client could provide different registration metadata for some reason) the same for all consumers of that `client` (e.g. https://app.element.io) by means of the IdP seeing the same registration metadata and issuing the same `client_id`

Author: Indeed, we shouldn't use the `client_id` as a device ID. The long-term goal is to de-duplicate client IDs, so that we can clearly state to users if they used this client in the past or not, and if other people are using this client on their homerserver.

The port number, path and query components MAY be different.
For example, if the `client_uri` is `https://example.com/`, then one of the `redirect_uris` can be `https://example.com/callback` or `https://app.example.com/callback`, but not `https://app.com/callback`.

#### User-visible metadata values

The following metadata values SHOULD be used by clients to help users identify the client:

 - `client_name`: Human-readable name of the client to be presented to the user
 - `logo_uri`: URL that references a logo for the client
 - `tos_uri`: URL that points to a human-readable terms of service document for the client
 - `policy_uri`: URL that points to a human-readable policy document for the client

All the URIs MUST use the `https` scheme and use the `client_uri` as a common base, as defined by the previous section.

If provided by the client, the homeserver SHOULD show or link to the `tos_uri` and `policy_uri` to the user.
They MUST NOT use a user or password in the authority component of the URI.
They MUST point to a valid web page and SHOULD NOT require authentication to access.

> All of these metadata values are optional.
> 
> As per [RFC 7591 sec. 2.2](https://tools.ietf.org/html/rfc7591#section-2.2), these metadata values MAY be localized.

Reviewer 8: ```suggestion
All of these metadata values are optional.

##### Metadata localization

As per [RFC 7591 sec. 2.2](https://tools.ietf.org/html/rfc7591#section-2.2), these metadata values MAY be localized.
```

Readability nit: as this feels like it should be a separate section.

With the previous registration request, the server would reply with:

```json
{
  "client_id": "s6BhdRkqt3",
  "client_name": "My App",
  "client_uri": "https://example.com/",
  "logo_uri": "https://example.com/logo.png",
  "tos_uri": "https://example.com/tos.html",
  "policy_uri": "https://example.com/policy.html",
  "redirect_uris": ["https://app.example.com/callback"],
  "token_endpoint_auth_method": "none",
  "response_types": ["code"],
  "grant_types": ["authorization_code", "refresh_token"],
  "application_type": "web"
}
```

**Note**: in this example, the server has not registered the locale-specific values for `client_name`, `tos_uri`, and `policy_uri`, which is why they are not present in the response. The server also does not support the `urn:ietf:params:oauth:grant-type:token-exchange` grant type, which is why it is not present in the response.

The client MUST store the `client_id` for future use.
It SHOULD reuse the `client_id` for all future authorization requests done against the same homeserver.

## Potential issues

> Because each client on each user device will do its own registration, they will all have different `client_id`s.

Reviewer 8: ```suggestion
Because each client on each user device will do its own registration, they may all have different `client_id`s.
```

as de-duplicating is possible in implementation.

# MSC2966: Usage of OAuth 2.0 Dynamic Client Registration in Matrix

This proposal is part of the broader [MSC3861: Next-generation auth for Matrix, based on OAuth 2.0/OIDC](https://github.com/matrix-org/matrix-spec-proposals/pull/3861).

> This MSC specifies how Matrix clients SHOULD leverage the OAuth 2.0 Dynamic Client Registration Protocol ([RFC 7591](https://tools.ietf.org/html/rfc7591)) to register themselves before initiating an authorization flow.

Reviewer 10: Not being familiar with the RFC I'm having a lot of trouble figuring out which parts are just "follow the MSC" and which parts are specific to Matrix.

Author: I've added two notes about this here: dbb61639

In short, nothing is Matrix-specific other than the set of metadata values we enforce/encourage clients to use

> In brief, once a client has obtained the homeserver's OAuth 2.0 metadata per [MSC2965](https://github.com/matrix-org/matrix-spec-proposals/pull/2965), and before it can initiate an authorization request per [MSC2964](https://github.com/matrix-org/matrix-spec-proposals/pull/2964), the client must register itself with that homeserver to obtain a `client_id`.

Reviewer 10: Probably a dumb question, but does *every* client have to do this? If I'm running a corporate server can I pre-register my-web-client.acme.com?

I'm not sure if that even makes sense.

Author: Because this is not in the actual proposal text, I will rephrase this to make it less authoritative, as this is already covered by 'Proposal > Dynamic client registration'. Did that in 55433db4

  "client_id": "s6BhdRkqt3",
  "client_name": "My App",
  "client_uri": "https://example.com/",
  "logo_uri": "https://example.com/logo.png",
  "tos_uri": "https://example.com/tos.html",
  "policy_uri": "https://example.com/policy.html",
  "redirect_uris": ["https://app.example.com/callback"],
  "token_endpoint_auth_method": "none",
  "response_types": ["code"],
  "grant_types": ["authorization_code", "refresh_token"],
  "application_type": "web"
}
```

**Note**: in this example, the server has not registered the locale-specific values for `client_name`, `tos_uri`, and `policy_uri`, which is why they are not present in the response. The server also does not support the `urn:ietf:params:oauth:grant-type:token-exchange` grant type, which is why it is not present in the response.

The client MUST store the `client_id` for future use.
It SHOULD reuse the `client_id` for all future authorization requests done against the same homeserver.

## Potential issues

Because each client on each user device will do its own registration, they may all have different `client_id`s.
This means that the number of client registrations will grow over time with the number of devices and unique clients.

> This increase can be mitigated by de-duplicating client registrations that have identical metadata.

Reviewer 10: This is an implementation detail/suggestion to implementors I assume?

# MSC2966: Usage of OAuth 2.0 Dynamic Client Registration in Matrix

This proposal is part of the broader [MSC3861: Next-generation auth for Matrix, based on OAuth 2.0/OIDC][MSC3861].

> This MSC specifies how Matrix clients SHOULD leverage the OAuth 2.0 Dynamic Client Registration Protocol ([RFC 7591](https://tools.ietf.org/html/rfc7591)) to register themselves before initiating an authorization flow.

Reviewer 11: Not to be a stickler, but I think having this 'should' in formal capitals technically implies the whole MSC is optional, which is probably not what you want.

# MSC2966: Usage of OAuth 2.0 Dynamic Client Registration in Matrix

[MSC2964](https://github.com/matrix-org/matrix-doc/pull/2964) defines how client should login against a Matrix server using OAuth 2.0.
It assumes the client is known to the authentication server.

This MSC specifies how Matrix clients should leverage OAuth 2.0 Dynamic Client Registration Protocol ([RFC 7591](https://tools.ietf.org/html/rfc7591)) to register themselves before initiating the login flow.

## Proposal

If a Matrix server wants to be used by any third-party client, its authentication server must allow dynamic registration of OAuth 2.0 clients.

> The client registration endpoint is advertised in the OIDC discovery document and can be used as per [RFC 7591 sec. 3](https://tools.ietf.org/html/rfc7591#section-3).

Reviewer 3: I'm sure this is obvious if you're familiar with OAuth2, but for context I'd find it really helpful if you could clarify *when* a client would register itself with a given server in the context of Matrix.  Presumably, whenever the user chooses a homeserver to log into? (Is the client expected to remember a list of homeservers it has registered against?)

Reviewer: (considering thread as resolved - if this feels untrue, please open a new thread)

| ---------------- | ----------------------------------------------------------------------------------------------------------- | ------------------------------------------------ | ----------- |
| `client_name`    | Human-readable name of the client to be presented to the user                                               | Required                                         | Yes         |
| `client_uri`     | URL of a web page providing information about the client                                                    | Optional                                         | Yes         |
| `logo_uri`       | URL that references a logo for the client                                                                   | Optional                                         | Yes         |
| `contacts`       | Array of strings representing ways to contact people responsible for this client, typically email addresses | Required                                         | No          |
| `tos_uri`        | URL that points to a human-readable terms of service document for the client                                | Required                                         | Yes         |

> | `policy_uri`     | URL that points to a human-readable policy document for the client                                          | Required                                         | Yes         |

Reviewer 12: Does this mean, that for any client I develop, I now need infrastructure and domains to host these documents, as I assume these need to be web accessible? Currently you don't need any of that, since the client doesn't send any data to the client developer and doesn't provide any service, where ToS make any sense. But now to be able to sign in with my self written client, I need to host a website? That doesn't sound like "Simpler for Homeservers and clients to implement" from MSC3861 at all.

Reviewer 13: Also often community clients are deliberately NOT a service provider. So why would they need a ToS if there is no Service?

For example Apache2 comes with this disclaimer in the license:

> Unless required by applicable law or agreed to in writing, software
> distributed under the License is distributed on an "AS IS" BASIS,
> WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
> See the License for the specific language governing permissions and
> limitations under the License.

So clearly there is no service provided.

Author: This requirement has been lifted, they are not required anymore

> | `redirect_uris`  | Array of redirection URIs for use in redirect-based flows                                                   | Required with the `authorization_code` grant ype | No          |

Reviewer 3: something I don't really understand here (or about OAuth2 dynamic registration in general):

What is to stop me, as an attacker, registering a "client" with branding (name, logo, etc) that looks like a trusted client, but with a redirect_uri that points to a site I control?

I then send the user a link which will take them through the login flow, and redirect back to my site. Assuming I can persaude the user to open the link, I can then harvest the authorization code and thence gain access to their account?

Author: In this case the auth server would at least prompt for explicit consent as it would be considered as a separate client from the legitimate one.

One thing that could be enforced would be to have the different URIs (at least `redirect_uris` and `client_uri`) to be on the same origin. Quoting from [RFC 7591 sec. 5](https://tools.ietf.org/html/rfc7591#section-5):

> In a situation where the authorization server is supporting open client registration, it must be extremely careful with any URL provided by the client that will be displayed to the user (e.g., `logo_uri`, `tos_uri`, `client_uri`, and `policy_uri`).  For instance, a rogue client could specify a registration request with a reference to a drive-by download in the `policy_uri`, enticing the user to click on it during the authorization.  The authorization server SHOULD check to see if the `logo_uri`, `tos_uri`, `client_uri`, and `policy_uri` have the same host and scheme as the those defined in the array of `redirect_uris` and that all of these URIs resolve to valid web pages.

Reviewer 3: > In this case the auth server would at least prompt for explicit consent as it would be considered as a separate client from the legitimate one.

of course; but how is the user meant to tell that they are now being asked for consent to a phishing app?

> One thing that could be enforced would be to have the different URIs (at least redirect_uris and client_uri) to be on the same origin.

Wouldn't this preclude the use of desktop clients, which often register custom URI schemes to get the auth code back from a browser to the client?

Reviewer 3: As discussed elsewhere: it turns out that on Windows and MacOS at least, desktop apps *can* register for parts of the `https://` namespace. So that only really leaves Linux where you can only register for an entire URI scheme (via an [`x-scheme-handler/...` mimetype](https://specifications.freedesktop.org/shared-mime-info-spec/shared-mime-info-spec-latest.html#idm45094474779648)).

So it should be possible to create a system which makes this look user-friendly for everyone except Linux users, who presumably are a bit more comfortable with obscure URIs anyway.

Reviewer 12: My client generates a localhost address on demand during SSO currently. This MSC makes it sound like I need to switch to a fixed schema instead, which means running multiple instances of my software in parallel will become much harder. How would I solve having 2 separate instances of my app running and I want to only sign into one of them?

Author: > My client generates a localhost address on demand during SSO currently. This MSC makes it sound like I need to switch to a fixed schema instead, which means running multiple instances of my software in parallel will become much harder. How would I solve having 2 separate instances of my app running and I want to only sign into one of them?

This is covered by the "Native clients" part of this spec: apps can register `localhost` without port, and the server must accept a dynamic port during the authorisation

Reviewer: (considering thread as resolved - if this feels untrue, please open a new thread)

| `response_types` | Array of the OAuth 2.0 response types that the client may use                                               | Defaults to `["code"]`                           | No          |
| `grant_types`    | Array of OAuth 2.0 grant types that the client may use                                                      | Defaults to `["authorization_code"]`             | No          |

Other metadata registered in the IANA [OAuth Dynamic Client Registration Metadata](https://www.iana.org/assignments/oauth-parameters/oauth-parameters.xhtml#client-metadata) registry might be used and supported by Matrix servers and clients.

### User consent

When authorizing a third-party client for the first time, the authorization server must ask for explicit user consent and display as much information, preferably localized, about the client as possible.
This includes informations about the publisher, the clients terms of service and its policy.

The consent screen must include a human-readable list of the scopes requested by the client.

### Metadata signature

To securely identify themselves, clients must send a digitally signed version of their metadata.

> This is done by encoding the client metadata in a JWT and signing it.

Reviewer 12: How do I include such data in my open-source project? Because this would allow either everyone to make themselves look like my client or I can't distribute my software freely anymore, can I? Is there some documentation on why this is secure for desktop applications, that I could read? Or what "securely identify" means in this context as in what it is protecting against?

Author: I've removed the signed metadata part of the MSC

> This is done by encoding the client metadata in a JWT and signing it.

Reviewer 4: It is not said _how_ the JWT is sent during registration. Looking at RFC7591, my guess is this what is called a software statement, but it should be said clearly.

Author: Removed the metadata signature from the MSC, will be addressed in a separate one

The issuer does not have to be an actual authorization server, but its metadata can include human-readable informations about the issuer.
Those informations can be displayed on the user consent screen to tell the user about the client's publisher.

The JWT payload may also include a `software_version` claim denoting the version of the client being registered.
This field should be treated as an opaque string by the authorization server.

A Matrix server may choose to only allow signed client to be registered.
It might also have an list of trusted issuers for software statements and only allow those to restrict third-party clients.

## Potential issues

It is unclear how metadata updates should be handled.
If a client changes its `redirect_uris`, should the old ones be considered for a period of time?
If multiple versions of the same client coexist at the same time, should older versions of the software be able to override the metadata of the newer version?

> If an authorization server allows unsigned clients, the number of client registration might explode exponentially.

Reviewer 12: Why is this a problem?

Reviewer 13: I don't think this should be problem, imho. This is the whole goal of an open ecosystem like matrix. Anyone at any time can build a client and use it. You are not vendor bound. Allowing this to be restricted imho goes against the core idea of an open ecosystem which matrix tries to be.

Reviewer: (considering thread as resolved - if this feels untrue, please open a new thread)

**Note**: in this example, the server has not registered the locale-specific values for `client_name`, `tos_uri`, and `policy_uri`, which is why they are not present in the response. The server also does not support the `urn:ietf:params:oauth:grant-type:token-exchange` grant type, which is why it is not present in the response.

The client MUST store the `client_id` for future use.

To avoid the number of client registrations growing over time, the server MAY choose to delete client registrations that don't have an active session.
The server MUST NOT delete client registrations that have an active session.

> Clients SHOULD perform a new client registration at the start of each authorization flow.

Reviewer 5: I'd probably say this is a `MUST`?

Author: done in ee36393

Reviewer 5: From out-of-band discussion: there may be valid reasons to not reregister, e.g. "like if the same client has multiple sessions with the same homeserver in parallel?". However, I'm paranoid that client implementers will see the SHOULD, end up caching the client ID and see that it works 90% of the time, and then things break if the server tries to delete the client ID mapping.

Suggestion is to convert to MUST, and add a caveat for that case, to make things explicit.
